File: 算法/001_two sum.java
// 1
import java.util.HashMap;
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int len = nums.length;
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++)
        {
        	if (map.containsKey(target - nums[i]))
        	{
        		return new int[] {map.get(target - nums[i]), i};
        	}
        	map.put(nums[i], i);
        }
        return null;
    }
}

class Solution {
	public int[] twoSum(int[] nums, int target){
		int len = nums.length;
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
        	final Integer value = map.get(nums[i]);
        	if (value != null) {
        		return new int[] {value, i};
        	}
        	map.put(target - nums[i], i);
        }
        return null;
    }
}
    


// 2
class Solution{
	public int[] twoSum(int[] nums, int target){
		for (int i = 0; i < nums.length; i++)
		{
			for (int j = i + 1; j < nums.length; i++)
			{
				if (nums[i] + nums[j] == target)
				{
					return new int[] {i, j};
				}
			}
		}
		return null;
	}
}

File: 算法/001_two sum.py
# 返回两个索引，使得其对应的值之和等于目标值，且唯一解
from typing import List
class Solution():

    def twoSum(self, nums: List[int], target: int) -> List[int]:
        dict = {num: pos for pos, num in enumerate(nums)}
        for pos, num in enumerate(nums):
            other = target - num
            other_pos = dict.get(other, -1)

            if other in dict and other_pos != pos:
                return [pos, other_pos]

        return [-1, -1]


    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashmap = {}
        for index, num in enumerate(nums):
            temp = target - num
            try:
                hashmap[temp]
                return [hashmap[temp], index]
            except KeyError:
                hashmap[num] = index
                
# brutal force O(n^2)
# 双指针
    def twoSum(self, nums, target):
        nums_index = [(item, index) for index, item in enumerate(nums)]
        nums_index.sort()
        begin, end = 0, len(nums)-1
        while begin < end:
            curr = nums_index[begin][0] + nums_index[end][0]
            if curr == target:
                return [nums_index[begin][1], nums_index[end][1]]
            elif curr < target:
                begin += 1
            else:
                end -= 1

#   def twoSum(self, nums, target):
#       length = len(nums)
#       for i in range(length):
#           for j in range(i+1, length):
#               if nums[i] + nums[j] == target:
#                   return [i, j]

#   def twoSum(self, nums, target):
#       if len(nums) == 0:
#           return []
#       for index, num in enumerate(nums):
#           for count in range(index+1, len(nums)):
#               if num + nums[count] == target:
#                   return [index, count]              
s = Solution()
nums = [2, 5, 8, 9]
print(s.twoSum(nums, 14))


# # 统计文件行数1
# count = len(open(filepath, 'r').readlines())
# # 统计文件行数2
# count = 0
# for index, line in enumerate(open(filepath, 'r')):
#     count += 1
